File: add_account.php
#! /usr/bin/env php

<?php

require_once("panoseti.inc");

function add_account_old($name, $passwd) {
    $passwd_path = '/home/wei/Projects/web/passwd.json';
    $passwds = json_decode(file_get_contents($passwd_path));
    if (isset($passwds->$name)) {
        die("An account with that name already exists\n");
    }
    $u = new stdClass;
    $u->auth =  bin2hex(random_bytes(16));
    $u->passwd_hash = hash_passwd($passwd);
    $passwds->$name = $u;
    file_put_contents($passwd_path, json_encode($passwds, JSON_PRETTY_PRINT));

    echo "Added account with name '$name' and passwd '$passwd'\n";
}

function add_account($name, $passwd) {
    $passwd_path = '/home/wei/Projects/web/passwd.json';

    if (file_exists($passwd_path)) {
        $json = file_get_contents($passwd_path);
        $passwds = json_decode($json);
        if (!is_object($passwds)) {
            $passwds = new stdClass();
        }
    } else {
        $passwds = new stdClass();
    }

    if (isset($passwds->$name)) {
        echo "Error: An account with name '$name' already exists.\n";
        return false;
    }

    $u = new stdClass();
    $u->auth = bin2hex(random_bytes(16));
    $u->passwd_hash = hash_passwd($passwd);

    $passwds->$name = $u;

    $json_data = json_encode($passwds, JSON_PRETTY_PRINT);
    if (file_put_contents($passwd_path, $json_data) === false) {
        echo "Error: Failed to write to '$passwd_path'.\n";
        return false;
    }

    echo "✅ Added account: '$name'\n";
    return true;
}

$stdin = fopen('php://stdin', 'r');
echo "Create a PanoSETI account\n";
echo "User name: ";
$name = trim(fgets($stdin));
echo "Password: ";
$passwd = trim(fgets($stdin));

add_account($name, $passwd);

?>

File: add_account.py
#! /usr/bin/env python3
import secrets
import hashlib
import json
import os

PASSWD_FILE = 'passwd.json'

def hash_passwd(p):
    combined = p + "blahblah"
    return hashlib.md5(combined.encode('utf-8')).hexdigest()

def generate_auth():
    return secrets.token_hex(16)

def load_accounts():
    if not os.path.exists(PASSWD_FILE):
        return {}
    try:
        with open(PASSWD_FILE, 'r') as f:
            return json.load(f)
    except:
        return {}
    
def save_accounts(accounts):
    with open(PASSWD_FILE, 'w') as f:
        json.dump(accounts, f, indent=4)

def main():
    username = input("Enter username: ").strip()
    password = input("Enter password: ").strip()

    accounts = load_accounts()

    if username in accounts:
        print(f"❌ An account with name '{username}' already exists.")
        return

    accounts[username] = {
        "auth": generate_auth(),
        "passwd_hash": hash_passwd(password)
    }

    save_accounts(accounts)
    print(f"✅ Account '{username}' added successfully.")

if __name__ == '__main__':
    main()